packages/yarnpkg-pnpify/sources/RawHoister.ts

```diff
@@ -0,0 +1,244 @@
+    const treeCopy: PackageTree = new Map();
+
+    for (const [nodeId, depIds] of tree)
+      treeCopy.set(nodeId, new Set(depIds));
+
+    return treeCopy;
+  }
+
+  /**
+   * Weighs all the packages in the subtree, by adding up own package weight with weights of all
+   * of its direct and transitive dependencies.
+   *
+   * @param tree package tree
+   * @param rootId root package id of the subtree
+   * @param packageMap package map
+   * @param nohoist nohoist package ids, that shouldn't be weighed
+   *
+   * @return map of package weights: package id -> total weight
+   */
+  private weighPackages(tree: PackageTree, rootId: PackageId, packageMap: PackageMap, nohoist: Set<PackageId>): WeightMap {
+    const weights: WeightMap = new Map();
+
+    const addUpNodeWeight = (nodeId: PackageId) => {
+      if (!nohoist.has(nodeId)) {
+        weights.set(nodeId, packageMap.get(nodeId)!.weight + (weights.get(nodeId) || 0));
+        for (const depId of tree.get(nodeId) || NO_DEPS) {
+          addUpNodeWeight(depId);
+        }
+      }
+    };
+
+    addUpNodeWeight(rootId);
+
+    return weights;
+  }
+
+  /**
+   * Finds packages that have the max weight among the packages with the same name
+   *
+   * @param weights package weights map: package id -> total weight
+   * @param packageMap package info map: package id -> package name
+   *
+   * @returns package ids with max weights among the packages with the same name
+   */
+  private getHeaviestPackages(weights: WeightMap, packageMap: PackageMap): Set<PackageId> {
+    const heaviestPackages = new Map<PackageName, {weight: Weight, pkgId: PackageId}>();
+    for (const [pkgId, weight] of weights) {
+      const pkgName = packageMap.get(pkgId)!.name;
+      let heaviestPkg = heaviestPackages.get(pkgName)!;
+      if (!heaviestPkg) {
+        heaviestPkg = {weight, pkgId};
+        heaviestPackages.set(pkgName, heaviestPkg);
+      } else if (weight > heaviestPkg.weight) {
+        heaviestPkg.weight = weight;
+        heaviestPkg.pkgId = pkgId;
+      }
+    }
+
+    const heavyPackageIds = new Set<PackageId>();
+    for (const [,pkg] of heaviestPackages)
+      heavyPackageIds.add(pkg.pkgId);
+
+    return heavyPackageIds;
+  }
+
+  /**
+   * Find the packages that can be hoisted to the subtree root `rootId`.
+   *
+   * @param tree package tree
+   * @param rootId package id that should be regarded as subtree root
+   * @param packageMap package map
+   * @param nohoist nohoist package ids
+   */
+  private computeHoistCandidates(tree: PackageTree, rootId: PackageId, packageMap: PackageMap, nohoist: Set<PackageId>): Set<PackageId> {
+    // Get current package dependency package names
+    const curDepNames = new Map<PackageName, PackageId>();
+    for (const depId of tree.get(rootId) || NO_DEPS)
+      curDepNames.set(packageMap.get(depId)!.name, depId);
+
+    // At first all the packages in the subtree are hoist candidates, we weigh them all
```

Comment:
I'm not sure this is right - for example in the following case:

```
. → A → B@X → C → B@Y
```

You don't want `B@Y` to be candidate for hoisting at the top-level, since it would cause `C` to require `B@X` instead.

Reply:
Good point. I can hoist only the packages from subtree, if they have been spotted only once on a path from the subtree root.